Panic in propertiesRelatedTo for recursive variadic tuple constraint

## Stack trace

Tested with `@typescript/native-preview` version
`7.0.0-dev.20260707.2`.

```text
panic: runtime error: index out of range [-1] [recovered, repanicked]

github.com/microsoft/typescript-go/internal/checker.(*Relater).propertiesRelatedTo
    internal/checker/relater.go:4199

github.com/microsoft/typescript-go/internal/checker.(*Relater).structuredTypeRelatedToWorker
    internal/checker/relater.go:3864

github.com/microsoft/typescript-go/internal/checker.(*Relater).recursiveTypeRelatedTo
    internal/checker/relater.go:3135
```

## Steps to reproduce

Create `repro.ts`:

```ts
type Inner<A extends any[], A2 extends [...A, boolean]> = A2;

export type Wrapper<A extends any[]> = Inner<
  A,
  [...A, boolean, boolean, boolean, boolean]
>;
```

Run:

```sh
npx -y -p @typescript/native-preview@latest tsgo --noEmit repro.ts
```

The compiler panics in `propertiesRelatedTo`.

## Expected behavior

The compiler should report any applicable diagnostic without panicking.

## Related work

- Legacy issue: https://github.com/microsoft/TypeScript/issues/62974
- Unmerged JavaScript implementation: https://github.com/microsoft/TypeScript/pull/62983

The earlier PR passed its test runs but was automatically closed when ordinary
development moved from the JavaScript implementation to `typescript-go`.

--- a/internal/checker/relater.go
+++ b/internal/checker/relater.go
@@ -4115,7 +4115,8 @@ func (r *Relater) propertiesRelatedTo(source *Type, target *Type, reportErrors b
 			} else {
 				sourceRest = true
 			}
-			targetHasRestElement := target.TargetTupleType().combinedFlags&ElementFlagsVariable != 0
+			targetHasRestElement := target.TargetTupleType().combinedFlags&ElementFlagsRest != 0
+			targetHasVariableElement := target.TargetTupleType().combinedFlags&ElementFlagsVariable != 0
 			var sourceMinLength int
 			if isTupleType(source) {
 				sourceMinLength = source.TargetTupleType().minLength
@@ -4129,13 +4130,13 @@ func (r *Relater) propertiesRelatedTo(source *Type, target *Type, reportErrors b
 				}
 				return TernaryFalse
 			}
-			if !targetHasRestElement && targetArity < sourceMinLength {
+			if !targetHasVariableElement && targetArity < sourceMinLength {
 				if reportErrors {
 					r.reportError(diagnostics.Source_has_0_element_s_but_target_allows_only_1, sourceMinLength, targetArity)
 				}
 				return TernaryFalse
 			}
-			if !targetHasRestElement && (sourceRest || targetArity < sourceArity) {
+			if !targetHasVariableElement && (sourceRest || targetArity < sourceArity) {
 				if reportErrors {
 					if sourceMinLength < targetMinLength {
 						r.reportError(diagnostics.Target_requires_0_element_s_but_source_may_have_fewer, targetMinLength)
@@ -4162,6 +4163,12 @@ func (r *Relater) propertiesRelatedTo(source *Type, target *Type, reportErrors b
 				if targetHasRestElement && sourcePosition >= targetStartCount {
 					targetPosition = targetArity - 1 - min(sourcePositionFromEnd, targetEndCount)
 				} else {
+					if sourcePosition >= targetArity {
+						if reportErrors {
+							r.reportError(diagnostics.Target_allows_only_0_element_s_but_source_may_have_more, targetArity)
+						}
+						return TernaryFalse
+					}
 					targetPosition = sourcePosition
 				}
 				targetFlags := ElementFlagsNone

--- a/testdata/baselines/reference/compiler/variadicTupleMismatch.errors.txt
+++ b/testdata/baselines/reference/compiler/variadicTupleMismatch.errors.txt
@@ -1,0 +1,32 @@
+variadicTupleMismatch.ts(4,5): error TS2322: Type '[...A, number, number, number, number]' is not assignable to type '[...A, number]'.
+  Target allows only 2 element(s) but source may have more.
+variadicTupleMismatch.ts(8,5): error TS2322: Type '[...A, number, number, number, number]' is not assignable to type '[number, ...A]'.
+  Variadic element at position 0 in source does not match element at position 0 in target.
+variadicTupleMismatch.ts(12,5): error TS2322: Type '[...A, number, number, number, number]' is not assignable to type '[...A, ...A]'.
+  Source provides no match for variadic element at position 1 in target.
+
+
+==== variadicTupleMismatch.ts (3 errors) ====
+    // https://github.com/microsoft/typescript-go/issues/4727
+    
+    function f1<A extends unknown[]>(x: [...A, number], y: [...A, number, number, number, number]) {
+        x = y  // Error
+        ~
+!!! error TS2322: Type '[...A, number, number, number, number]' is not assignable to type '[...A, number]'.
+!!! error TS2322:   Target allows only 2 element(s) but source may have more.
+    }
+    
+    function f2<A extends unknown[]>(x: [number, ...A], y: [...A, number, number, number, number]) {
+        x = y  // Error
+        ~
+!!! error TS2322: Type '[...A, number, number, number, number]' is not assignable to type '[number, ...A]'.
+!!! error TS2322:   Variadic element at position 0 in source does not match element at position 0 in target.
+    }
+    
+    function f3<A extends unknown[]>(x: [...A, ...A], y: [...A, number, number, number, number]) {
+        x = y  // Error
+        ~
+!!! error TS2322: Type '[...A, number, number, number, number]' is not assignable to type '[...A, ...A]'.
+!!! error TS2322:   Source provides no match for variadic element at position 1 in target.
+    }
+    

--- a/testdata/baselines/reference/compiler/variadicTupleMismatch.symbols
+++ b/testdata/baselines/reference/compiler/variadicTupleMismatch.symbols
@@ -1,0 +1,45 @@
+//// [tests/cases/compiler/variadicTupleMismatch.ts] ////
+
+=== variadicTupleMismatch.ts ===
+// https://github.com/microsoft/typescript-go/issues/4727
+
+function f1<A extends unknown[]>(x: [...A, number], y: [...A, number, number, number, number]) {
+>f1 : Symbol(f1, Decl(variadicTupleMismatch.ts, 0, 0))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 2, 12))
+>x : Symbol(x, Decl(variadicTupleMismatch.ts, 2, 33))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 2, 12))
+>y : Symbol(y, Decl(variadicTupleMismatch.ts, 2, 51))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 2, 12))
+
+    x = y  // Error
+>x : Symbol(x, Decl(variadicTupleMismatch.ts, 2, 33))
+>y : Symbol(y, Decl(variadicTupleMismatch.ts, 2, 51))
+}
+
+function f2<A extends unknown[]>(x: [number, ...A], y: [...A, number, number, number, number]) {
+>f2 : Symbol(f2, Decl(variadicTupleMismatch.ts, 4, 1))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 6, 12))
+>x : Symbol(x, Decl(variadicTupleMismatch.ts, 6, 33))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 6, 12))
+>y : Symbol(y, Decl(variadicTupleMismatch.ts, 6, 51))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 6, 12))
+
+    x = y  // Error
+>x : Symbol(x, Decl(variadicTupleMismatch.ts, 6, 33))
+>y : Symbol(y, Decl(variadicTupleMismatch.ts, 6, 51))
+}
+
+function f3<A extends unknown[]>(x: [...A, ...A], y: [...A, number, number, number, number]) {
+>f3 : Symbol(f3, Decl(variadicTupleMismatch.ts, 8, 1))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 10, 12))
+>x : Symbol(x, Decl(variadicTupleMismatch.ts, 10, 33))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 10, 12))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 10, 12))
+>y : Symbol(y, Decl(variadicTupleMismatch.ts, 10, 49))
+>A : Symbol(A, Decl(variadicTupleMismatch.ts, 10, 12))
+
+    x = y  // Error
+>x : Symbol(x, Decl(variadicTupleMismatch.ts, 10, 33))
+>y : Symbol(y, Decl(variadicTupleMismatch.ts, 10, 49))
+}
+

--- a/testdata/baselines/reference/compiler/variadicTupleMismatch.types
+++ b/testdata/baselines/reference/compiler/variadicTupleMismatch.types
@@ -1,0 +1,38 @@
+//// [tests/cases/compiler/variadicTupleMismatch.ts] ////
+
+=== variadicTupleMismatch.ts ===
+// https://github.com/microsoft/typescript-go/issues/4727
+
+function f1<A extends unknown[]>(x: [...A, number], y: [...A, number, number, number, number]) {
+>f1 : <A extends unknown[]>(x: [...A, number], y: [...A, number, number, number, number]) => void
+>x : [...A, number]
+>y : [...A, number, number, number, number]
+
+    x = y  // Error
+>x = y : [...A, number, number, number, number]
+>x : [...A, number]
+>y : [...A, number, number, number, number]
+}
+
+function f2<A extends unknown[]>(x: [number, ...A], y: [...A, number, number, number, number]) {
+>f2 : <A extends unknown[]>(x: [number, ...A], y: [...A, number, number, number, number]) => void
+>x : [number, ...A]
+>y : [...A, number, number, number, number]
+
+    x = y  // Error
+>x = y : [...A, number, number, number, number]
+>x : [number, ...A]
+>y : [...A, number, number, number, number]
+}
+
+function f3<A extends unknown[]>(x: [...A, ...A], y: [...A, number, number, number, number]) {
+>f3 : <A extends unknown[]>(x: [...A, ...A], y: [...A, number, number, number, number]) => void
+>x : [...A, ...A]
+>y : [...A, number, number, number, number]
+
+    x = y  // Error
+>x = y : [...A, number, number, number, number]
+>x : [...A, ...A]
+>y : [...A, number, number, number, number]
+}
+

--- a/testdata/tests/cases/compiler/variadicTupleMismatch.ts
+++ b/testdata/tests/cases/compiler/variadicTupleMismatch.ts
@@ -1,0 +1,15 @@
+// @noEmit: true
+
+// https://github.com/microsoft/typescript-go/issues/4727
+
+function f1<A extends unknown[]>(x: [...A, number], y: [...A, number, number, number, number]) {
+    x = y  // Error
+}
+
+function f2<A extends unknown[]>(x: [number, ...A], y: [...A, number, number, number, number]) {
+    x = y  // Error
+}
+
+function f3<A extends unknown[]>(x: [...A, ...A], y: [...A, number, number, number, number]) {
+    x = y  // Error
+}

